### PR TITLE
L3 post-fork zero-copy host buffers

### LINKED ISSUE #1027: L3 worker cannot see host tensors created after worker startup
## Summary

`DistributedWorker` / L3 worker currently requires all host input/output tensors passed to `worker.run(...)` to be created and moved to shared memory before the L3 worker child process is created. If a later run uses a newly-created tensor, or a tensor with a shape/storage that was not shared before worker creation, the child process cannot reliably see that storage.

This creates a practical limitation for dynamic-shape serving workloads: every possible input/output buffer shape must be preallocated up front, otherwise repeated `worker.run(...)` calls with changed input shapes may fail or require awkward fixed-size buffer workarounds.

## Context

This was observed while migrating `pypto-serving` Qwen3-14B prefill/decode dispatch from the L2 worker path to a shared L3 `DistributedWorker` path:

- Feature issue: https://github.com/hw-native-sys/pypto-serving/issues/26
- Implementation PR: https://github.com/hw-native-sys/pypto-serving/pull/29

In that PR, serving currently has to preallocate fixed-size host buffers for prefill/decode before creating the L3 worker:

- prefill hidden / seq_lens / chunk_lens / chunk_offsets / block_table / slot_mapping / logits
- decode hidden / seq_lens / block_table / slot_mapping / logits

The worker is then created once and reused for both prefill and decode. This works, but only because all runtime input/output buffers have stable shared-memory storage before the child process starts.

## Problem

For dynamic serving workloads, input shapes can change between runs:

- prefill total token count changes per request batch / chunk
- user batch size changes
- block table and slot mapping sizes may vary with max sequence / page layout
- output logits buffers may vary with batch shape in other models

With the current L3 worker behavior, tensors created after the worker child exists are not visible to that child unless they were already backed by shared memory visible at fork/spawn time. This means a natural pattern like:

```python
worker = DistributedWorker([program])

for request_batch in batches:
    hidden = torch.empty((dynamic_tokens, hidden_size), dtype=torch.bfloat16)
    out = torch.empty((dynamic_batch, vocab), dtype=torch.float32)
    worker.run(program, hidden, out)
```

is not safe unless `hidden` and `out` are replaced with slices of pre-created shared-memory buffers.

## Current workaround

`pypto-serving` now preallocates maximum-capacity buffers before L3 worker creation and reuses slices for each run. For example:

- prefill uses a fixed `kernel_batch * max_seq` hidden buffer but passes only the active token slice
- fixed-size metadata buffers are zeroed / filled each run
- decode uses fixed `kernel_batch` buffers and pads inactive rows

This workaround is functional, but it has drawbacks:

- higher memory footprint because maximum-size buffers must exist up front
- model runner code becomes more complex
- dynamic-shape programs need fixed-capacity host buffers even when only a small shape is active
- shape changes are constrained by preallocated capacity
- child-process visibility details leak into serving-side code

## Expected behavior

The L3 worker runtime should support one of the following:

1. Passing newly-created shared-memory host tensors to an already-running L3 worker, with the child process able to attach/map them for that run.
2. A documented runtime API to register or refresh host tensor storage mappings after worker creation.
3. A clear error if a non-visible host tensor is passed to `worker.run(...)`, instead of failing indirectly or forcing users to discover the preallocation requirement.
4. Documentation explaining the exact lifetime/visibility rules for host tensor arguments used by `DistributedWorker.run(...)`.

## Why this matters

Serving engines naturally have dynamic request batches. Requiring every possible host input/output tensor to be created before worker startup makes the L3 worker path harder to use as a general replacement for the L2 runtime path.

The limitation is manageable for Qwen3-14B today through fixed-size buffers, but it becomes fragile for other models, variable vocab/logit outputs, dynamic prefill chunking, or future serving paths that need more flexible input shapes.

## Validation from pypto-serving

After applying the fixed preallocation workaround in `pypto-serving`, these checks pass:

- offline generation without larger PTO2 runtime envs
- single-request HTTP serving without larger PTO2 runtime envs
- multi-request HTTP serving with:

```text
PTO2_RING_HEAP=4294967296
PTO2_RING_TASK_WINDOW=1048576
PTO2_RING_DEP_POOL=1048576
```

This confirms the workaround is viable, but the underlying L3 child-process tensor visibility limitation remains.

### PR DESCRIPTION
**Pull request by doraemonmj**
Host tensors created *after* the L3 chip children are forked (lazily on the
first `run()`) were invisible to those children, forcing serving to
preallocate every buffer before worker creation. This adds
`Worker.create_host_buffer` / `free_host_buffer`: a born-shared named shm is
mapped into every chip child, and the user builds a tensor over its
`memoryview` (`torch.frombuffer` / `np.frombuffer`), so the child reads and
writes the same physical pages the parent sees — no host-side per-run copy to
make the memory child-visible. The per-task mailbox blob's host pointers are
rewritten to the child's own mapping before the runtime dereferences them.

Pure Python (`worker.py` / `orchestrator.py`) — no runtime C++ change; the
simpler core stays framework-free (torch/numpy appear only on the caller's
side). `submit_next_level` validates each in-range host tensor fits its buffer
(raising on overrun) and forwards fork-inherited tensors unvalidated.

- Lifetime/visibility contract documented in `docs/comm-domain.md`.
- One a2a3sim scene test covers the round-trip; host-side staging is unit-tested.

## Performance

Measured fork (`share_memory_`) vs create on a2a3, 2 chips. `create_host_buffer`
is **per-run neutral**, not a speedup. Its cost profile relative to the fork
path:

- **per-run**: equal within noise (`simpler_run` fork ≈ 7466 µs / create ≈
  7714 µs). The device H2D is registration-agnostic, so the steady-state
  `run()` does not change.
- **only net recurring create cost**: a µs-scale per-submit host-pointer rewrite
  (`_rewrite_blob_host_addrs`) — create maps the shm at a different child VA, so
  in-range pointers are redirected every submit; fork inherits the same VA and
  skips it (`if host_buf_ranges:` is false). Scales with tensor count, below the
  noise floor.
- **~720 ms one-time `_start_hierarchical`** (lazy chip fork + per-rank device
  init + callable pre-warm) is L3 startup **both** paths pay once — fork at the
  first `run()`, create at the first `create_host_buffer` — so it is a
  trigger-point difference, not a create-specific overhead.
- **buffer provisioning** (excl. that one-time startup) is ~equal (16 MB: fork
  59.6 / create 54.7 ms), dominated by memory first-touch both paths pay; the
  interfaces themselves are cheap and size-flat.

So create trades a µs-per-submit rewrite for the capability fork cannot offer —
a child-visible host buffer allocated *after* the chips are forked.

Closes #1027

### CODE REVIEW

File: python/simpler/worker.py [hunk 1]
```diff
@@ -1139,6 +1320,11 @@ def _run_chip_main_loop(  # noqa: PLR0912, PLR0913, PLR0915 -- unified TASK_READ
                 control_shm.close()
             except Exception:  # noqa: BLE001
                 pass
+        for host_shm, _lo, _hi, _base in host_buf_table.values():
+            try:
+                host_shm.close()
+            except Exception:  # noqa: BLE001
+                pass
```
[the remark this answers is not in the document]
Comment (doraemonmj) [reply python/simpler/worker.py:1327]: Declining this one. S110 (flake8-bandit `S`) is **not** in this repo's ruff `select` (`pyproject.toml` selects only `PL, I, E, W, F, UP`), so repo CI does not flag it — the `# noqa: BLE001` is the only suppression the code needs. This `host_shm.close()` teardown block is also intentionally identical in shape to the sibling `control_shm.close()` block directly above it (same silent best-effort `except Exception: pass`); adding a log to only one would break that local consistency. Keeping as-is.

File: python/simpler/worker.py [hunk 2]
```diff
@@ -3786,6 +3986,279 @@ def copy_from(self, dst: int, src: int, size: int, worker_id: int = 0) -> None:
...
+                try:
+                    shm.unlink()
+                except FileNotFoundError:
+                    pass
+            raise
+
+        buf_view = shm.buf
+        assert buf_view is not None
+        return HostBuffer(token=token, data_ptr=data_ptr, nbytes=nbytes, buffer=buf_view)
+
```
[the remark this answers is not in the document]
Comment (doraemonmj) [reply python/simpler/worker.py:4101]: Good catch — adopted in 36ad8f9a. Wrapped the rollback `_broadcast_host_unmap(token)` in `except Exception` and log to stderr, so an unmap failure during rollback can no longer propagate in place of the original error (the trailing `raise` now always re-raises the real cause, e.g. "MAP_HOST failed on N chip children"). This matches the catch-and-log pattern `free_host_buffer` already uses.

File: python/simpler/worker.py [hunk 3]
```diff
@@ -3786,6 +3986,279 @@ def copy_from(self, dst: int, src: int, size: int, worker_id: int = 0) -> None:
...
+                if self._worker is not None and getattr(self, "_hierarchical_started", False):
+                    self._broadcast_host_unmap(entry.token)
+            except Exception:  # noqa: BLE001
+                # A failed unmap broadcast must not strand the parent shm.
+                pass
+            finally:
+                # Tolerates a still-live view over a zero-copy buffer at close():
+                # unlinks the name regardless so the OS reclaims it once dropped.
+                self._close_host_shm(entry)
+
```
[the remark this answers is not in the document]
Comment (doraemonmj) [reply python/simpler/worker.py:4184]: Adopted in 36ad8f9a — but for diagnostic symmetry, not S110 (which this repo does not select). The point stands: `free_host_buffer` already logs its UNMAP_HOST failures while this close-path swallowed them silently. Now both log to stderr with the token and a "continuing best-effort" note.

File: tests/st/a2a3/tensormap_and_ringbuffer/test_l3_host_buffer_registration.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            a.fill_(5.0)  # in place → lands directly in the child-visible shm
+            b.fill_(7.0)
+            out.zero_()
+            worker.run(_one_task_orch(chip_handle, a, b, out), args=None, config=CallConfig())
+            assert torch.allclose(out, _golden(a, b), rtol=self.RTOL, atol=self.ATOL)
+            del a, b, out  # drop the views so free_host_buffer can release the shm
+        finally:
+            worker.free_host_buffer(ba)
+            worker.free_host_buffer(bb)
+            worker.free_host_buffer(bout)
```
[the remark this answers is not in the document]
Comment (doraemonmj) [reply tests/st/a2a3/tensormap_and_ringbuffer/test_l3_host_buffer_registration.py:137]: Restructured in 36ad8f9a: `a = b = out = None` up front, `del` moved into `finally` before the frees, and the `torch.allclose` result asserted after the `finally`. So cleanup of all three buffers runs even if `run()`/the assertion fails.

One correction on the stated mechanism though: freeing with a live view does **not** raise here. `_close_host_shm` catches the `BufferError` from `shm.close()` and returns a warning string (it does not re-raise), and `free_host_buffer` only writes that to stderr — `unlink` runs regardless. So on the old code an assertion failure was not actually masked and `bb`/`bout` were not leaked (the OS reclaims on process exit). The restructure is still worth it as hardening + it drops the 3× `BufferError` warn noise on the failure path.

**Review by ChaoZheng109 — APPROVED** [review]
